### PR TITLE
Glutin update

### PR DESCRIPTION
**Pull request by paulrouget**
[Review on Reviewable](https://reviewable.io/reviews/servo/glutin/43)

### CODE REVIEW

**Comment by paulrouget** [comment]
It requires https://github.com/servo/servo/pull/7096

**Comment by paulrouget** [comment]
This is basically glutin/master +

@glennw	Fix a rare crash in some X11 implementations (details below). 	1d90924
@larsbergstrom	Add X11 features required for linking against ffi functions	68f5399
@metajack	Send travis notifications to Homu	c1b6971
@zmike	add back headless feature	4fb60f1
@zmike	add ability to set parent window when creating windows	 9f7669f

@glennw mentioned that 1d90924 should land upstream.
I believe zmike's commits are also importants.

Can someone confirm the 2 other commits are also required?

And if anyone knows why it's crashing…

File: src/api/x11/window.rs [hunk 1]
```diff
@@ -0,0 +1,814 @@
...
+                message_type: 0,
+                serial: 0,
+                send_event: 0,
+                display: data.display,
+                data: unsafe { mem::zeroed() },
+            };
+
+            unsafe {
+                ffi::XSendEvent(data.display, data.window, 0, 0, mem::transmute(&mut xev));
+                ffi::XFlush(data.display);
```
Comment (tomaka) [inline src/api/x11/window.rs]: Must be `(self.x.display.xlib.XSendEvent)` and `(self.x.display.xlib.XFlush)`

**Comment by tomaka** [comment]
I don't know how it compiles for you (travis also seems to agree that it shouldn't compile).

Your error seems to indicate that `#version` is missing in one of the example codes. If this is the case, it's a mistake in the example.

One of the latest changes in glutin is that it uses EGL by default on Linux instead of GLX (it still uses OpenGL and not OpenGL ES, and falls back to GLX if EGL doesn't support OpenGL). It is possible that the context given by EGL is stricter when enforcing errors.

**Comment by paulrouget** [comment]
I'm on osx.

That crash is in rust-layers. Not reference to `#version` there. Does it mean I need to add `#version` to all the shaders?

**Comment by paulrouget** [comment]
@tomaka Adding `#version 100` helped.

See https://github.com/paulrouget/rust-layers/commit/27a4085e25fb0be7f0f8287daf96b0191c7153f9

Now I'm getting this message:

```
thread '<main>' panicked at 'Failed to compile shader: ERROR: 0:11: 'sampler2DRect' : declaration must include a precision qualifier for type
```

Using `sampler2D` instead doesn't panic, but then I get a grey screen (servo, osx).

Can you explain a bit what's going on? I don't know much about opengl, any pointer would help.

Looking at Linux now…

**Comment by tomaka** [comment]
Try `#version 110` instead.

**Comment by tomaka** [comment]
Does it work for you before the glutin update?

OS/X only supports shader versions 100 or 140+. If you want version 100 you need to add precision qualifiers to all the variables `mediump`. If you want version 140 then the `attribute` and `varying` keywords no longer exist.
So I don't see how rust-layers even works in its current state.

**Comment by tomaka** [comment]
Oh, wait, duh.

I said that OS/X didn't support version 110, but that's only true if you use a recent version of OpenGL.

There's also the possibility to create a legacy context (3.1 or less) at initialization, in which case it works.

This is what servo is doing:
 - https://github.com/servo/servo/blob/44c4bb00c1cb8645ee2fc303848a5136108e594f/ports/glutin/window.rs#L141-L144
 - https://github.com/servo/servo/blob/44c4bb00c1cb8645ee2fc303848a5136108e594f/ports/glutin/window.rs#L96

**Comment by metajack** [comment]
@mrobinson @glennw What should we be doing here? Should we upgrade the context or keep using the legacy one?

**Comment by paulrouget** [comment]
> @mrobinson @glennw What should we be doing here? Should we upgrade the context or keep using the legacy one?

For the record, using `NSOpenGLProfileVersionLegacy` works.

If that's ok, we can use the legacy context and once the update has landed, we upgrade?

**Comment by metajack** [comment]
I think it's fine to keep doing what we were doing and upgrade later if that is the correct thing long term.

**Comment by paulrouget** [comment]
Note: we hit a Rust bug in the `match` code that selects the right OpenGL profile. Not using ranges appears to solve the issue.

**Comment by paulrouget** [comment]
See https://github.com/tomaka/glutin/pull/567 for profile selection fix.

**Comment by paulrouget** [comment]
@glennw / @tomaka  can you please tell me if both of these patches are needed, and where they should go (servo/glutin or tomaka/glutin):

* https://github.com/tomaka/glutin/commit/5124bda848810605208e488790e379978a7fcb22 ("This was previously in glutin, but was removed since Weak is not marked stable here: 0f7bd90")
* https://github.com/glennw/glutin/commit/7ee0318ba648a6012db43c1f367eee24753ec0d6

And does one require another?

7ee0318ba648a6012db43c1f367eee24753ec0d6 is a bit tricky because I don't know what to do in `wakeup_event_loop` where things changed quite a bit (`display.display` and `x11.display`).

**Comment by metajack** [comment]
I believe Glenn took out Weak<T> support so https://github.com/tomaka/glutin/commit/5124bda848810605208e488790e379978a7fcb22 is no longer needed for current Servo.

https://github.com/glennw/glutin/commit/7ee0318ba648a6012db43c1f367eee24753ec0d6 is needed though. But seems like this could go upstream. @tomaka ?

**Comment by tomaka** [comment]
Sure. As long as it's not unstable I'm ok with it.

**Comment by paulrouget** [comment]
Upstream PRs:
* https://github.com/tomaka/glutin/pull/567
* https://github.com/tomaka/glutin/pull/568

**Comment by paulrouget** [comment]
All required PRs have now been merged upstreams.
Can someone review this?

**Comment by paulrouget** [comment]
Once merged, I'll update servo's Cargo.lock in https://github.com/servo/servo/pull/7096

**Comment by paulrouget** [comment]
Review status: 0 of 83 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

<sup>**[src/api/x11/window.rs, line 122 \[r1\]](https://reviewable.io:443/reviews/servo/glutin/43#-JwCXnkdMDCiX2MGbkJR-r1-122)** ([raw file](https://github.com/servo/glutin/blob/9f7669f338adb5896c73a76f1e6a8f3b8b30813c/src/api/x11/window.rs#L122)):</sup>
Addressed. Change made it upstream.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/glutin/43)

**Comment by glennw** [comment]
@paulrouget This is quite difficult to review as is. Perhaps we could rename the existing servo branch to servo_<date> and create a new servo branch from glutin/master then just apply our local patches on top? (We've done this in the past a few times). It would be a lot easier to see the differences and give us a cleaner history.

**Comment by paulrouget** [comment]
This PR is against the master branch, not the servo branch. Creating a new PR: #44

**Pull request closed without merging**